Fix error 500 caused by ActionController::RoutingError (via fail-safe) when POST parameters are invalid

weppos opened the pull request:
### Summary

This PR fixes #29983 that I reported a few hours go.

In summary, when a request for an invalid route (that triggers a `ActionController::RoutingError`) contains invalid POST parameters (that can't be parsed), Rails crashes when trying to display the error and it returns an error 500 (because of the fail-safe mechanism) instead of a 404.

The ticket #29983 contains all the details, along with the backtraces of the bug I found in the app logs.

### Other Information

This PR includes a test to reproduce the error and avoid regressions. In #29983 I originally suggested to relax the rescue to StandardError, but instead I decided to add the specific exceptions in the rescue statement.

I will be happy to use a StandardError if you prefer.

weppos committed:
Reproduce error caused by malformed parameters

    Error:
    RequestFormat#test_format_does_not_throw_exceptions_when_invalid_POST_parameters:
    ActionDispatch::Http::Parameters::ParseError: 765: unexpected token at '{record:{content:24.12.1.146}}'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/parameters.rb:113:in `rescue in parse_formatted_parameters'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/parameters.rb:107:in `parse_formatted_parameters'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/request.rb:360:in `block in POST'
        /Users/weppos/.rvm/gems/ruby-2.3.3/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch'
        /Users/weppos/.rvm/gems/ruby-2.3.3/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch_header'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/request.rb:359:in `POST'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/parameters.rb:53:in `parameters'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:62:in `block in formats'
        /Users/weppos/.rvm/gems/ruby-2.3.3/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch'
        /Users/weppos/.rvm/gems/ruby-2.3.3/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch_header'
        /Users/weppos/Mirrors/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:60:in `formats'
        /Users/weppos/Mirrors/rails/actionpack/test/dispatch/request_test.rb:891:in `block in <class:RequestFormat>'

See GH-29983

weppos committed:
Capture parameter parsing error output and test it

This change prevents the log to be displayed in the tests.
Moreover, the assertion against the debug ensures that
the test effectively triggers the parsing error as expected.

weppos committed:
Use a generic value in the test

gmcgibbon commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -60,7 +60,7 @@ def formats
           params_readable = begin
                               parameters[:format]
-                            rescue ActionController::BadRequest
+                            rescue ActionController::BadRequest, Parameters::ParseError
```
Maybe we should put it somewhere in here instead?

https://github.com/rails/rails/blob/98c1432583d4607af1467425183d8d448a692c5a/actionpack/lib/action_dispatch/http/parameters.rb#L54-L58

weppos replied:
@gmcgibbon you knows it better than me. Feel free to adjust the PR.

gmcgibbon commented on actionpack/test/dispatch/request_test.rb:
```diff
@@ -874,12 +874,28 @@ class RequestFormat < BaseRequestTest
+    assert request.format.html?
+
+    output.rewind && (err = output.read)
+    assert err =~ /Error occurred while parsing request parameters/
```
You can do `assert_match` here instead for better failures.

weppos replied:
✅  changed

gmcgibbon reviewed (commented):
Still happens on `6.0.0.alpha`. Also worth noting this doesn't happen in development mode, only production. What environment setting adds mime negotiation?

weppos committed:
Switch to assert_match

weppos commented:
@gmcgibbon I applied the suggested changes. There's one left, which may be beyond my knowledge of the implementation.

Any chance you can consider this PR for 6.x? It has been open for a while and we keep having a significant amount of logs generated.

rafaelfranca commented:
Needs a rebase, but looks good to me. Can you rebase it?

weppos committed:
Rebase against master

The issue was fixed in 6b3faf8e502dbd238fe4b4c409e96308638297a1 (#34341).
Leaving the spec for coverage.

weppos commented:
@rafaelfranca thanks! I've rebased with master.

Actually, it looks like the bug was fixed in 6b3faf8e502dbd238fe4b4c409e96308638297a1, when the PR #34341 was merged. There was a change to use `RESCUABLE_MIME_FORMAT_ERRORS` that now includes this exception too.

I've merged the code anyways, and the PR is now basically just a test that protects against a regression.

Feel free to merge it, or close it if you don't think it's necessary.
